README.md
# spectr

[![check-deploy](https://github.com/hugheylab/spectr/workflows/check-deploy/badge.svg)](https://github.com/hugheylab/spectr/actions)
[![codecov](https://codecov.io/gh/hugheylab/spectr/branch/master/graph/badge.svg)](https://codecov.io/gh/hugheylab/spectr)
[![Netlify Status](https://api.netlify.com/api/v1/badges/eea83c4a-3e36-4245-8195-d0c15115832b/deploy-status)](https://app.netlify.com/sites/kind-edison-268732/deploys)
[![CRAN Status](https://www.r-pkg.org/badges/version/spectr)](https://cran.r-project.org/package=spectr)

`spectr` provides a consistent interface to use various methods to calculate the periodogram of a time-course. The methods include Lomb-Scargle, fast Fourier transform, and three versions of the chi-square periodogram. To see how we've used `spectr`, check out the [paper](https://doi.org/10.1371/journal.pcbi.1008567).

## Installation

### Option 1: CRAN

```r
install.packages('spectr')
```

### Option 2: Hughey Lab Drat Repository

1. Install [`BiocManager`](https://cran.r-project.org/package=BiocManager).

    ```r
    if (!requireNamespace('BiocManager', quietly = TRUE))
      install.packages('BiocManager')
    ```

1. If you use RStudio, go to Tools → Global Options... → Packages → Add... (under Secondary repositories), then enter:

    - Name: hugheylab
    - Url: https://hugheylab.github.io/drat/

    You only have to do this once. Then you can install or update the package by entering:

    ```r
    BiocManager::install('spectr')
    ```

    Alternatively, you can install or update the package by entering:

    ```r
    BiocManager::install('spectr', site_repository = 'https://hugheylab.github.io/drat/')
    ```

## Usage

See the examples and detailed guidance in the [reference documentation](https://spectr.hugheylab.org/reference/index.html).


NEWS.md
# spectr 1.0.2
* Match styling to lab standard.

# spectr 1.0.1
* Revised code to not need `globalVariables()` in order to pass R CMD check.

# spectr 1.0.0
* Updated for lomb 2.0

# spectr 0.0.0.9017
* Switched to using `stats::na.fail` for default `na.action`

# spectr 0.0.0.9016
* Moved example in documentation to roxygen comments to avoid funny business in devtools testing

# spectr 0.0.0.9015
* Removed `pval` from output of `cspgram`, since `log_pval` is preferred
* Simplified testing suite

# spectr 0.0.0.9014
* Fixed setting of attribute in `cspgram` so `data.table` doesn't throw a warning
* Added `pkgdown` site


tests/testthat.R
library(testthat)
library(spectr)

test_check('spectr')


tests/testthat/test_pgram.R
set.seed(1789)
deltat = 0.1
tau = 25
tt = seq(0, 24 * 3, deltat)
x = 3 * sin(tt / tau * 2 * pi) + rnorm(length(tt))

periodRange = c(20, 30)
dopar = FALSE

test_that('CSP greedy', {
  name = 'csp_greedy'
  specObs = cspgram(x, deltat, periodRange = periodRange,
                    method = 'greedy', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('CSP conservative', {
  name = 'csp_conservative'
  specObs = cspgram(x, deltat, periodRange = periodRange,
                    method = 'conservative', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('CSP standard', {
  name = 'csp_standard'
  specObs = cspgram(x, deltat, periodRange = periodRange,
                    method = 'standard', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('FFT', {
  name = 'fft'
  specObs = fftpgram(x, deltat, periodRange = periodRange)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('Lomb-Scargle', {
  name = 'lsp'
  specObs = lspgram(x, deltat, periodRange = periodRange)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})


tests/testthat/test_spectr.R
set.seed(1789)
deltat = 0.1
tau = 25
tt = seq(0, 24 * 3, deltat)
x = 3 * sin(tt / tau * 2 * pi) + rnorm(length(tt))

periodRange = c(20, 30)
dopar = FALSE

test_that('spectr CSP greedy', {
  name = 'spectr_greedy'
  specObs = spectr(x, deltat, periodRange = periodRange,
                   method = 'greedy_chisq', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('spectr CSP conservative', {
  name = 'spectr_conservative'
  specObs = spectr(x, deltat, periodRange = periodRange,
                   method = 'conservative_chisq', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('spectr CSP standard', {
  name = 'spectr_standard'
  specObs = spectr(x, deltat, periodRange = periodRange,
                   method = 'standard_chisq', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('spectr Lomb-Scargle', {
  name = 'spectr_lsp'
  specObs = spectr(x, deltat, periodRange = periodRange,
                   method = 'lombscargle', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})

test_that('spectr FFT', {
  name = 'spectr_fft'
  specObs = spectr(x, deltat, periodRange = periodRange,
                   method = 'fft', dopar = dopar)
  # qs::qsave(specObs, sprintf('spec_%s.qs', name))
  specExp = qs::qread(sprintf('spec_%s.qs', name))
  expect_equal(specObs, specExp)
})


R/cspgram.R
#' @rdname spectr
#' @export
cspgram = function(x, deltat, periodRange = c(18, 32),
                   method = c('greedy', 'conservative', 'standard'),
                   na.action = stats::na.fail, dopar = FALSE) {
  p = period = df = log_pval = chisq = NULL

  method = match.arg(method)
  checkX(x)
  checkSingleNum(deltat, 0, FALSE)
  checkPeriodRange(periodRange)

  pSpan = (periodRange[1L] / deltat):(periodRange[2L] / deltat)
  if (length(pSpan) == 0) {
    stop('deltat and periodRange are incompatible.')}

  if (isTRUE(dopar)) {
    doOp = `%dopar%`
  } else {
    doOp = `%do%`}

  x = na.action(x)
  k = length(x) %/% max(pSpan) # for conservative
  m = mean(x) # for greedy
  n = length(x)

  d = doOp(foreach(p = pSpan, .combine = rbind), {
    if (method == 'greedy') {
      k = n / p
      xNow = c(x, rep.int(NA, ceiling(k) * p - n))
      xMat = matrix(xNow, ncol = p, byrow = TRUE)
      xH = colMeans(xMat, na.rm = TRUE)
      qP = k * n * sum((xH - m)^2) / sum((x - m)^2)
    } else {
      if (method == 'standard') {
        k = n %/% p}
      xNow = x[1:(k * p)]
      mNow = mean(xNow)
      xMat = matrix(xNow, ncol = p, byrow = TRUE)
      xH = colMeans(xMat)
      qP = k * length(xNow) * sum((xH - mNow)^2) / sum((xNow - mNow)^2)}
    dNow = data.table(p = p, chisq = qP)})

  d[, period := p * deltat]
  d[, df := p - 1]
  d[, p := NULL]

  d[, log_pval := stats::pchisq(chisq, df, lower.tail = FALSE, log.p = TRUE)]

  data.table::setcolorder(d, 'period')
  data.table::setattr(d, 'method', method)
  return(d[])}


R/check_inputs.R
checkX = function(x) {
  stopifnot(is.vector(x, 'numeric'))
  invisible(0)}


checkSingleNum = function(z, lower, inclusive) {
  if (isTRUE(inclusive)) {
    op = `>=`
  } else {
    op = `>`}
  stopifnot(is.vector(z, 'numeric'),
            length(z) == 1L,
            op(z, lower))
  invisible(0)}


checkTime = function(x, deltat, time) {
  stopifnot(!(missing(deltat) && missing(time)))

  if (missing(time)) {
    checkSingleNum(deltat, 0, FALSE)
    time = seq(0, by = deltat, length.out = length(x))
  } else {
    stopifnot(is.vector(time, 'numeric'),
              length(time) == length(x))}
  return(time)}


checkPeriodRange = function(periodRange) {
  stopifnot(is.vector(periodRange, 'numeric'),
            length(periodRange) == 2L,
            all(periodRange >= 0),
            diff(periodRange) > 0)
  invisible(0)}


R/cspgram_example.R
library('data.table')
library('testthat')

set.seed(1789)
deltat = 0.1
tau = 25
tt = seq(0, 24 * 3, deltat)
x = 3 * sin(tt / tau * 2 * pi) + rnorm(length(tt))
cspGr = cspgram(x, deltat,periodRange = c(20,30), method = 'greedy', dopar = FALSE)

write.csv(cspGr, file = "cspGr.csv")

cspGrPeak = cspGr[which.min(log_pval)]

cspGrPeakExp = data.table(period = 24.4,
                          chisq = 635.4408,
                          df = 243,
                          log_pval = (-83.23465),
                          pval = (7.106419 * 10 ^ (-37)))
expect_true(all.equal(cspGrPeak,cspGrPeakExp,check.attributes=FALSE))


cspSt = cspgram(x, deltat, method = 'standard', dopar = FALSE)
write.csv(cspSt, file = "cspSt.csv")
cspStPeak = cspSt[which.min(log_pval)]

cspStPeakExp = data.table(period = 24,
                          chisq = 616.0034,
                          df = 239,
                          log_pval = (-79.13604),
                          pval = (4.282085 * 10 ^ (-35)))
expect_true(all.equal(cspStPeak,cspStPeakExp,check.attributes=FALSE,tolerance = 0.001))


cspCo = cspgram(x, deltat, method = 'conservative', dopar = FALSE)
write.csv(cspCo, file = "cspCo.csv")
cspCoPeak = cspCo[which.min(log_pval)]

cspCoPeakExp = data.table(period = 25.5,
                          chisq = 468.7745,
                          df = 254,
                          log_pval = (-32.75673),
                          pval = (5.942029 * 10 ^ (-15)))
expect_true(all.equal(cspCoPeak,cspCoPeakExp,check.attributes=FALSE,tolerance = 0.001))


R/spectr.R
#' @importFrom data.table data.table as.data.table := %between%
#' @importFrom foreach foreach %do% %dopar%
NULL




#' Calculate periodogram
#'
#' Calculate periodogram for a time-course using Lomb-Scargle, fast Fourier
#' transform, or selected version of chi-square. The `spectr` function is a
#' wrapper for the various methods. `lspgram` is in turn a wrapper for
#' [lomb::lsp()], and `fftpgram` a wrapper for [stats::spec.pgram()]. Among the
#' versions of chi-square, it is highly recommended to use greedy, which has
#' lower bias than standard and lower variance than conservative.
#'
#' @param x Numeric vector of measurements.
#' @param deltat Numeric value of the interval between time-points.
#' @param time Numeric vector of time-points. Can be specified instead of
#'   `deltat` for Lomb-Scargle.
#' @param periodRange Numeric vector of the minimum and maximum values of the
#'   period to consider, in the same units as `deltat` or `time`.
#' @param method Character indicating which method to use. Can be an unambiguous
#'   substring of the full name.
#' @param pad Numeric value of the proportion of the length of `x` by which to
#'   pad `x` with zeros. Must be > 0. Only used for FFT.
#' @param ofac Integer value of the oversampling factor. Must be >= 1. Only used
#'   for Lomb-Scargle.
#' @param na.action Function specifying how to handle `NA` values in `x`.
#'   Default is [stats::na.fail()], which gives an error if any values are
#'   missing. Ignored for Lomb-Scargle.
#' @param dopar Logical indicating whether to run calculations in parallel if
#'   a parallel backend is already set up, e.g., using
#'   [doParallel::registerDoParallel()]. Only used for chi-square.
#' @param ... Other arguments passed to [stats::spec.pgram()] for FFT.
#'
#' @return A `data.table` with various columns depending on the method. For any
#'   version of chi-square, columns will be `period`, `chisq`, `df`, and
#'   `log_pval`. The log p-value is more reliable than the p-value, since R has
#'   finite precision, so p-values less than about 5e-324 would be set to 0. For
#'   Lomb-Scargle and FFT, columns will be `period` and `power`.
#'
#' @examples
#' library('data.table')
#'
#' set.seed(1789)
#' deltat = 0.1
#' tau = 25
#' tt = seq(0, 24 * 3, deltat)
#' x = 3 * sin(tt / tau * 2 * pi) + rnorm(length(tt))
#'
#' specCsp = spectr(x, deltat, method = 'greedy')
#' peakCsp = specCsp[which.min(log_pval)]
#'
#' specLsp = spectr(x, deltat, method = 'lomb')
#' peakLsp = specLsp[which.max(power)]
#'
#' specFft = spectr(x, deltat, method = 'fft')
#' peakFft = specFft[which.max(power)]
#'
#' @export
spectr = function(
  x, deltat, time, periodRange = c(18, 32),
  method = c('greedy_chisq', 'conservative_chisq', 'standard_chisq',
             'lombscargle', 'fft'),
  ofac = 50, pad = 50, na.action = stats::na.fail, dopar = FALSE, ...) {

  method = match.arg(method)

  if (endsWith(method, 'chisq')) {
    spec = cspgram(
      x, deltat, periodRange, gsub('_chisq', '', method), na.action, dopar)
  } else if (method == 'lombscargle') {
    spec = lspgram(x, deltat, time, periodRange, ofac)
  } else { # method == 'fft'
    spec = fftpgram(x, deltat, periodRange, pad, na.action, ...)}
  return(spec)}
